### PR TITLE
Add AWS Secret/Auth

### PR DESCRIPTION
**Pull request by ltcarbonell**
Add benchmark tests for AWS Secret and Auth Engines

### CODE REVIEW

File: docs/tests/auth-aws.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+- `sts_region` `(string: "")` - Region to override the default region for making AWS STS API calls. Should only be set if `sts_endpoint` is set. If so, should be set to the region in which the custom `sts_endpoint` resides.
+- `iam_server_id_header_value` `(string: "")` - The value to require in the `X-Vault-AWS-IAM-Server-ID` header as part of GetCallerIdentity requests that are used in the iam auth method. If not set, then no value is required or validated. If set, clients must include an X-Vault-AWS-IAM-Server-ID header in the headers of login requests, and further this header must be among the signed headers validated by AWS. This is to protect against different types of replay attacks, for example a signed request sent to a dev server being resent to a production server. Consider setting this to the Vault server's DNS name.
+- `allowed_sts_header_values` `(string: "")` A comma separated list of additional request headers permitted when providing the iam_request_headers for an IAM based login call. In any case, a default list of headers AWS STS expects for a GetCallerIdentity are allowed.
+
+### AWS Role Configuration (`role`)`
```
Comment (elliesterner) [inline docs/tests/auth-aws.md:19]:
```suggestion
### AWS Role Configuration (`test_user`)`
```

File: benchmarktests/target_auth_aws.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	Endpoint               string   `hcl:"endpoint,optional"`
+	IAMEndpoint            string   `hcl:"iam_endpoint,optional"`
+	STSEndpoint            string   `hcl:"sts_endpoint,optional"`
+	STSRegion              string   `hcl:"sts_region,optional"`
+	IAMServerIDHeaderValue string   `hcl:"iam_server_id_header_value,optional"`
+	AllowedSTSHeaderValues []string `hcl:"allowed_sts_header_values,optional"`
+}
+
+type AWSTestUserConfig struct {
+	Role                       string `hcl:"role,optional"`
```
Comment (elliesterner) [inline benchmarktests/target_auth_aws.go:65]: should this be required?

File: benchmarktests/target_auth_aws.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	awsAuthUser, err := structToMap(a.config.AWSTestUserConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding aws auth user from struct: %v", err)
+	}
+
+	// Create AWS Test Role
+	setupLogger.Trace(writingLogMessage("aws auth user config"))
+	_, err = client.Logical().Write("auth/"+authPath+"/role/"+a.config.AWSTestUserConfig.Role, awsAuthUser)
+	// print out what we are writing
+	fmt.Print("auth/"+authPath+"/role/"+a.role, awsAuthUser)
```
Comment (elliesterner) [inline benchmarktests/target_auth_aws.go:199]: do we want to keep this?

File: docs/tests/secret-aws.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+## Test Parameters
+
+### AWS Database Configuration `connection`
+
+- `max_retries` `(int: -1)` - Number of max retries the client should use for
+  recoverable errors. The default (`-1`) falls back to the AWS SDK's default
+  behavior.
+- `access_key` `(string: <required>)` – Specifies the AWS access key ID.
+- `secret_key` `(string: <required>)` – Specifies the AWS secret access key.
```
Comment (elliesterner) [inline docs/tests/secret-aws.md:13]: let's mention the aws access and secret_key env vars :)

File: benchmarktests/target_secret_aws.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+// Main Config Struct
+type AWSSecretTestConfig struct {
+	AWSConnectionConfig *AWSConnectionConfig `hcl:"connection,block"`
+	AWSRoleConfig       *AWSRoleConfig       `hcl:"role,block"`
+}
+
+type AWSConnectionConfig struct {
+	MaxRetries       int    `hcl:"max_retries,optional"`
+	AccessKey        string `hcl:"access_key"`
+	SecretKey        string `hcl:"secret_key"`
```
Comment (elliesterner) [inline benchmarktests/target_secret_aws.go:52]: should these be optional if they can be passed in with env var

File: docs/tests/secret-aws.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  ```
+  {{ if (eq .Type "STS") }}
+      {{ printf "vault-%s-%s" (unix_time) (random 20) | truncate 32 }}
+  {{ else }}
+      {{ printf "vault-%s-%s-%s" (printf "%s-%s" (.DisplayName) (.PolicyName) | truncate 42) (unix_time) (random 20) | truncate 64 }}
+  {{ end }}
+  ```
+
+### Role Config
```
Comment (elliesterner) [inline docs/tests/secret-aws.md:37]:
```suggestion
### Role Config `role`
```

File: benchmarktests/target_secret_aws.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	Region           string `hcl:"region,optional"`
+	IAMEndpoint      string `hcl:"iam_endpoint,optional"`
+	STSEndpoint      string `hcl:"sts_endpoint,optional"`
+	UsernameTemplate string `hcl:"username_template,optional"`
+}
+
+type AWSRoleConfig struct {
+	Name                   string `hcl:"name,optional"`
+	CredentialType         string `hcl:"credential_type,optional"`
```
Comment (elliesterner) [inline benchmarktests/target_secret_aws.go:61]: `name` and `credential_type` are required in the markdown docs.

**Review by elliesterner — APPROVED** [review]
Looks great!!

File: docs/tests/auth-aws.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
+# AWS Authentication Credential Benchmark (`aws_auth`)
+
+This benchmark tests the performance of logins using the aws auth method.
```
Comment (davidadeleon) [inline docs/tests/auth-aws.md:3]:
```suggestion
This benchmark tests the performance of logins using the AWS auth method.
```

File: docs/tests/auth-jwt.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# Redis Static Credential Benchmark (`jwt_auth`)
+# JWT Static Credential Benchmark (`jwt_auth`)
 
 This benchmark tests the performance of logins using the jwt auth method.
```
Comment (davidadeleon) [inline docs/tests/auth-jwt.md:3]:
```suggestion
This benchmark tests the performance of logins using the JWT auth method.
```

**Review by davidadeleon — APPROVED** [review]
Few tiny things, but otherwise looks great!

File: benchmarktests/target_secret_aws.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+	return &AWSTest{
+		pathPrefix: "/v1/" + secretPath,
+		header:     generateHeader(client),
+		roleName:   a.config.AWSRoleConfig.Name,
+		logger:     a.logger,
+	}, nil
+}
+
+func (m *AWSTest) Flags(fs *flag.FlagSet) {}
```
Comment (davidadeleon) [inline benchmarktests/target_secret_aws.go:188]:
```suggestion
func (a *AWSTest) Flags(fs *flag.FlagSet) {}
```

**Pull request merged**